File: patterns/blinkstyle.go
package patterns

import (
	"time"

	"github.com/anderstorpsfestivalen/slisko/pkg/faker"
	"github.com/anderstorpsfestivalen/slisko/pkg/pixel"
	"github.com/anderstorpsfestivalen/slisko/pkg/traffic"
	"github.com/anderstorpsfestivalen/slisko/pkg/utils"
)

// BlinkStyle defines how a linecard's ports should blink
type BlinkStyle struct {
	// Timing parameters for the blink pattern (base values, scaled by traffic)
	MinInterval time.Duration
	MaxInterval time.Duration
	MinBlink    time.Duration
	MaxBlink    time.Duration

	// Blinker parameters
	MinBlinks float64
	MaxBlinks float64
	MinCycle  time.Duration
	MaxCycle  time.Duration

	// Color parameters for different speeds
	SlowColor ColorStyle
	FastColor ColorStyle
	DeadColor ColorStyle

	// Dead port probability (0.0-1.0, e.g., 0.067 for ~1 in 15)
	DeadPortChance float64

	// Slow speed probability (0.0-1.0, e.g., 0.2 for 20% slow)
	SlowSpeedChance float64
}

// ColorStyle defines RGB color values (0.0-1.0)
type ColorStyle struct {
	RMult float64
	GMult float64
	BMult float64
}

// PortState represents the state of a single port
type PortState struct {
	faker faker.Fake
	port  *pixel.Pixel
	style ColorStyle
	isDead bool
}

// CreatePort creates a new port with this blink style
func (bs *BlinkStyle) CreatePort(port *pixel.Pixel, shaper *traffic.Shaper) PortState {
	// Check if this should be a dead port
	if randFloat() < bs.DeadPortChance {
		return PortState{
			port:   port,
			style:  bs.DeadColor,
			isDead: true,
		}
	}

	// Determine speed (slow or fast)
	style := bs.FastColor
	if randFloat() < bs.SlowSpeedChance {
		style = bs.SlowColor
	}

	// Scale timing based on traffic
	minInterval, maxInterval := shaper.GetScaledInterval(bs.MinInterval, bs.MaxInterval)
	minBlink, maxBlink := shaper.GetScaledInterval(bs.MinBlink, bs.MaxBlink)

	return PortState{
		faker: faker.NewRandomInterval(
			minInterval,
			maxInterval,
			minBlink,
			maxBlink,
			faker.NewRandomBlinker(bs.MinBlinks, bs.MaxBlinks, bs.MinCycle, bs.MaxCycle),
		),
		port:   port,
		style:  style,
		isDead: false,
	}
}

// Render renders a single port's state
func (ps *PortState) Render() {
	if ps.isDead {
		ps.port.SetClamped(ps.style.RMult, ps.style.GMult, ps.style.BMult)
		return
	}

	v := utils.Invert(ps.faker.Trig())
	ps.port.SetClamped(v*ps.style.RMult, v*ps.style.GMult, v*ps.style.BMult)
}

// Helper function for random floats
func randFloat() float64 {
	return utils.Random(0.0, 1.0)
}

// Predefined blink styles for different linecard types

// ASR9000Style returns the blink style for ASR9000 40GE linecards
func ASR9000Style() BlinkStyle {
	return BlinkStyle{
		MinInterval:     50 * time.Millisecond,
		MaxInterval:     7 * time.Second,
		MinBlink:        50 * time.Millisecond,
		MaxBlink:        12 * time.Second,
		MinBlinks:       15,
		MaxBlinks:       40,
		MinCycle:        1 * time.Second,
		MaxCycle:        10 * time.Second,
		SlowColor:       ColorStyle{RMult: 1.0, GMult: 0.6, BMult: 0.0},
		FastColor:       ColorStyle{RMult: 0.3, GMult: 1.0, BMult: 0.0},
		DeadColor:       ColorStyle{RMult: 1.0, GMult: 0.0, BMult: 0.0},
		DeadPortChance:  0.067, // ~1 in 15
		SlowSpeedChance: 0.2,   // 20% slow
	}
}

// Cisco7609Style returns the blink style for Cisco 7609 linecards
func Cisco7609Style() BlinkStyle {
	return BlinkStyle{
		MinInterval:     100 * time.Millisecond,
		MaxInterval:     7 * time.Second,
		MinBlink:        100 * time.Millisecond,
		MaxBlink:        12 * time.Second,
		MinBlinks:       15,
		MaxBlinks:       40,
		MinCycle:        1 * time.Second,
		MaxCycle:        10 * time.Second,
		SlowColor:       ColorStyle{RMult: 1.0, GMult: 0.5, BMult: 0.0},
		FastColor:       ColorStyle{RMult: 0.3, GMult: 1.0, BMult: 0.0},
		DeadColor:       ColorStyle{RMult: 1.0, GMult: 0.0, BMult: 0.0},
		DeadPortChance:  0.0,   // No dead ports
		SlowSpeedChance: 0.2,   // 20% slow
	}
}

File: firmware/slisko-core/src/patterns/blinkstyle.rs
//! Ported from `patterns/blinkstyle.go`.
//!
//! A [`BlinkStyle`] describes how a linecard's ports blink; [`PortState`] is one
//! port's runtime state (a faker + target pixel index + color), built by
//! [`BlinkStyle::create_port`]. Durations are `f32` seconds (Go used
//! `time.Duration`).

use alloc::boxed::Box;

use crate::faker::{Fake, RandomBlinker, RandomInterval, Rng};
use crate::pattern::BootstrapCtx;
use crate::pixel::Pixel;
use crate::utils;

/// RGB multipliers (0.0..1.0), mirrors `ColorStyle`.
#[derive(Clone, Copy, Debug)]
pub struct ColorStyle {
    pub r: f32,
    pub g: f32,
    pub b: f32,
}

/// Mirrors `BlinkStyle`. Timing fields are seconds.
#[derive(Clone, Copy, Debug)]
pub struct BlinkStyle {
    pub min_interval: f32,
    pub max_interval: f32,
    pub min_blink: f32,
    pub max_blink: f32,
    pub min_blinks: f32,
    pub max_blinks: f32,
    pub min_cycle: f32,
    pub max_cycle: f32,
    pub slow_color: ColorStyle,
    pub fast_color: ColorStyle,
    pub dead_color: ColorStyle,
    pub dead_port_chance: f32,
    pub slow_speed_chance: f32,
}

/// One port's runtime state (mirrors `PortState`).
pub struct PortState {
    faker: Option<Box<dyn Fake + Send>>,
    port: usize,
    style: ColorStyle,
    is_dead: bool,
}

impl PortState {
    /// Render this port into the strand (mirrors `PortState.Render`).
    pub fn render(&mut self, leds: &mut [Pixel], now: f32) {
        let px = &mut leds[self.port];
        if self.is_dead {
            px.set_clamped(self.style.r, self.style.g, self.style.b);
            return;
        }
        let v = utils::invert(self.faker.as_mut().map(|f| f.trig(now)).unwrap_or(0.0));
        px.set_clamped(v * self.style.r, v * self.style.g, v * self.style.b);
    }
}

impl BlinkStyle {
    /// Build a [`PortState`] for `port` (mirrors `BlinkStyle.CreatePort`):
    /// dead-port and slow/fast decisions via the ctx RNG, timings scaled by the
    /// ctx intensity.
    pub fn create_port(&self, port: usize, ctx: &mut BootstrapCtx) -> PortState {
        if ctx.rng.range_f32(0.0, 1.0) < self.dead_port_chance {
            return PortState {
                faker: None,
                port,
                style: self.dead_color,
                is_dead: true,
            };
        }

        let style = if ctx.rng.range_f32(0.0, 1.0) < self.slow_speed_chance {
            self.slow_color
        } else {
            self.fast_color
        };

        let inv = 1.0 / ctx.intensity.max(0.1); // scaled_interval divides by intensity
        let blinker = RandomBlinker::new(
            self.min_blinks,
            self.max_blinks,
            self.min_cycle,
            self.max_cycle,
            0.0,
            Rng::new(ctx.rng.next_seed()),
        );
        let faker = RandomInterval::new(
            self.min_interval * inv,
            self.max_interval * inv,
            self.min_blink * inv,
            self.max_blink * inv,
            Box::new(blinker),
            0.0,
            Rng::new(ctx.rng.next_seed()),
        );

        PortState {
            faker: Some(Box::new(faker)),
            port,
            style,
            is_dead: false,
        }
    }
}

/// Blink style for ASR9000 40GE linecards (mirrors `ASR9000Style`).
pub fn asr9000_style() -> BlinkStyle {
    BlinkStyle {
        min_interval: 0.05,
        max_interval: 7.0,
        min_blink: 0.05,
        max_blink: 12.0,
        min_blinks: 15.0,
        max_blinks: 40.0,
        min_cycle: 1.0,
        max_cycle: 10.0,
        slow_color: ColorStyle {
            r: 1.0,
            g: 0.6,
            b: 0.0,
        },
        fast_color: ColorStyle {
            r: 0.3,
            g: 1.0,
            b: 0.0,
        },
        dead_color: ColorStyle {
            r: 1.0,
            g: 0.0,
            b: 0.0,
        },
        dead_port_chance: 0.067,
        slow_speed_chance: 0.2,
    }
}

/// Blink style for Cisco 7609 linecards (mirrors `Cisco7609Style`).
pub fn cisco7609_style() -> BlinkStyle {
    BlinkStyle {
        min_interval: 0.1,
        max_interval: 7.0,
        min_blink: 0.1,
        max_blink: 12.0,
        min_blinks: 15.0,
        max_blinks: 40.0,
        min_cycle: 1.0,
        max_cycle: 10.0,
        slow_color: ColorStyle {
            r: 1.0,
            g: 0.5,
            b: 0.0,
        },
        fast_color: ColorStyle {
            r: 0.3,
            g: 1.0,
            b: 0.0,
        },
        dead_color: ColorStyle {
            r: 1.0,
            g: 0.0,
            b: 0.0,
        },
        dead_port_chance: 0.0,
        slow_speed_chance: 0.2,
    }
}
